File: LeetCode_POTD/July_2026/July_11.py
"""
Create the graph adjacency list and then for each if have not been visited yet, traverse complete connected subgraph and check if all nodes were having (subgraph size)-1 edges to ensure completeness

TC - O(n), SC - O(n)
"""

class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        edg=[0]*n
        adj=[[] for _ in range(n)]
        for i,j in edges:
            adj[i].append(j)
            adj[j].append(i)
            edg[i]+=1
            edg[j]+=1
        ans=0
        l=deque()
        vis=[False]*n
        for i in range(n):
            if not vis[i]:
                flg=True
                cnt=edg[i]-1
                vis[i]=True
                for j in adj[i]: l.append(j)
                while l:
                    a=l.popleft()
                    if vis[a]: continue
                    vis[a]=True
                    if cnt<0 or edg[i]!=edg[a]: 
                        flg=False
                    cnt-=1
                    for j in adj[a]: l.append(j)
                if flg: ans+=1
        return ans

File: LeetCode_POTD/July_2026/July_11.java
package Leetcode_POTD.July_2026;

/*
Create the graph adjacency list and then for each if have not been visited yet, traverse complete connected subgraph and check if all nodes were having (subgraph size)-1 edges to ensure completeness

TC - O(n), SC - O(n)
*/

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Queue;

public class July_11 {
    public int countCompleteComponents(int n, int[][] edges) {
        int[] edg=new int[n];
        ArrayList<Integer>[] adj=new ArrayList[n];
        for(int i=0;i<n;i++) adj[i]=new ArrayList<>();
        for(int[] i:edges){
            adj[i[0]].add(i[1]);
            adj[i[1]].add(i[0]);
            edg[i[0]]++;
            edg[i[1]]++;
        }
        int ans=0;
        Queue<Integer> l=new ArrayDeque<>();
        boolean[] vis=new boolean[n];
        for(int i=0;i<n;i++){
            if(!vis[i]){
                boolean flg=true;
                int cnt=edg[i]-1;
                vis[i]=true;
                for(int j:adj[i]) l.offer(j);
                while(!l.isEmpty()){
                    int a=l.poll();
                    if(vis[a]) continue;
                    vis[a]=true; 
                    if(cnt--<0 || edg[i]!=edg[a]) flg=false;
                    for(int j:adj[a]) l.offer(j);
                }
                if(flg) ans++;
            }
        }
        return ans;
    }
}
